src/wrapper_array.jl

```diff
@@ -0,0 +1,115 @@
+# `L` is for labels, it should be a `Tuple` of `Symbol`s
+struct NamedDimsArray{L, T, N, A<:AbstractArray{T,N}} <: AbstractArray{T,N}
+    data::A
+end
+
+function NamedDimsArray{L}(orig::AbstractArray{T,N}) where {L, T, N}
+    if !(L isa NTuple{N, Symbol})
+        throw(ArgumentError(
+            "A $N dimentional array, needs a $N-tuple of dimension names. Got: $L"
+        ))
+    end
+    return NamedDimsArray{L, T, N, typeof(orig)}(orig)
+end
+function NamedDimsArray(orig::AbstractArray{T,N}, names::NTuple{N, Symbol}) where {T, N}
+    return NamedDimsArray{names}(orig)
+end
+
+parent_type(::Type{<:NamedDimsArray{L,T,N,A}}) where {L,T,N,A} = A
+Base.parent(x::NamedDimsArray) = x.data
+
+
+"""
+    dim_names(A)
+
+Returns a tuple of containing the names of all the dimensions of the array `A`.
+"""
+dim_names(::Type{<:NamedDimsArray{L}}) where L = L
+dim_names(x::T) where T<:NamedDimsArray = dim_names(T)
+
+
+name2dim(a::NamedDimsArray{L}, name) where L = name2dim(L, name)
+
+
+
+#############################
+# AbstractArray Interface
+# https://docs.julialang.org/en/v1/manual/interfaces/index.html#man-interface-array-1
+
+## Minimal
+Base.size(a::NamedDimsArray) = size(parent(a))
+Base.size(a::NamedDimsArray, dim) = size(parent(a), name2dim(a, dim))
+
+
+## optional
+Base.IndexStyle(::Type{A}) where A<:NamedDimsArray = Base.IndexStyle(parent_type(A))
+
+Base.length(a::NamedDimsArray) = length(parent(a))
+
+Base.axes(a::NamedDimsArray) = axes(parent(a))
+Base.axes(a::NamedDimsArray, dim) = axes(parent(a), name2dim(a, dim))
+
+
+function Base.similar(a::NamedDimsArray{L}, args...) where L
```

Comment:
This will only work when the returned array from `similar` has the same number of dimensions. Maybe the type signature can reflect this?

Reply:
I don't think so.
`similar` has *so many* overloads.
And we want to leave them the args alone.

Reply:
Based on looking at all the overloads in stdlib/Base, I think it's pretty easy to separate it into two categories: one with dims, and one without. And it would be simple to make this only apply without dims:
```suggestion
function Base.similar(a::NamedDimsArray{L}, args::Type...) where L
```

Reply:
Why the `...`s?
This can't match anything that is used except the type form.
Which is probably ok,
but in that case `...` isn't needed

Reply:
Some SparseArrays methods take two types. 

It appears the only thing that is ever used that isn't types is dims.

Reply:
Ok, cool